### DISCUSSION TITLE
Insert custom BinaryType values without the need of json::binary() ctor

### DISCUSSION THREAD

**Discussion by lemonpy** [nlohmann/json#4209] [Ideas]
Let's say that I have my basic_json type with my custom binary class, the following test fails
```
class Buffer : public std::vector<char>
{
...
};
using json = nlohmann::basic_json <
             std::map,
             std::vector,
             std::string,
             bool,
             std::int64_t,
             std::uint64_t,
             double,
             std::allocator,
             nlohmann::adl_serializer,
             Buffer /* binary_t */ >;

TEST_CASE("Use BinaryType assign ctor")
{
    json empty;
    Buffer buf(5, '1');
    empty["buf"] = buf;
    REQUIRE(empty["buf"].is_binary());
}
```
`empty["buf"]` results being an array.
Currently, you have to call `json::binary()` explicitly to insert a binary correctly no matter if you are inserting a BinaryType value into the json.

**Comment by nlohmann** [comment]
This looks like a real inconsistency: assigning a custom `BinaryType` value directly via `operator=` produces an array rather than a binary value, forcing an explicit `json::binary(buf)` call even though the type information is already known at compile time via the template parameter. Since custom `BinaryType` is an advanced, rarely-used customization point, this has likely gone unnoticed rather than being an intentional design choice, and would be worth fixing via a constructor/assignment overload that detects the configured `BinaryType`.

---
*This reply was drafted by [Claude Code](https://claude.com/claude-code) on behalf of @nlohmann, as part of a review of open discussions.*
